File: src-java/algoanalysis/FastTest.java
package algoanalysis;

import acm.program.ConsoleProgram;
import java.util.HashMap;
import java.util.Map;

public class FastTest extends ConsoleProgram {

    private static final int FACT_N = 20;
    private static final long NUM_ITERATIONS = 100000000;
    private Map<Integer, Long> lookupTable = new HashMap<>();
    private long[] lookupTableArray = new long[FACT_N + 1];

    public long iterativeFac(int n) {
        long product = 1;
        for (long k = 1; k <= n; k++)
            product *= k;
        return  product;
    }

    public long recursiveFac(int n) {
        if (n == 0)   // base case
            return 1;

        return n * recursiveFac(n - 1);   // recursive case
    }

    public long stirlingFac(int n) {
        // Approximation
        return Math.round(Math.sqrt(2 * Math.PI * n) * Math.pow(n / Math.E, n));
    }

    public long lookupTableFac(int n) {
        // Dynamic Programming
        if (!this.lookupTable.containsKey(n))
            this.lookupTable.put(n, iterativeFac(n));
        return this.lookupTable.get(n);
    }

    public long lookupTableArrayFac(int n) {
        // Dynamic Programming
        if (this.lookupTableArray[n] == 0)
            this.lookupTableArray[n] = iterativeFac(n);
        return this.lookupTableArray[n];
    }

    public void run() {
        setFont("Arial-bold-32");


        long fac = 0;

        long timeBefore = System.currentTimeMillis();
        for (int i = 0; i < NUM_ITERATIONS; i++)
            fac = iterativeFac(FACT_N);
        long duration = (System.currentTimeMillis() - timeBefore);
        println("Iterative: " + fac + " (" + duration + "ms)");

        timeBefore = System.currentTimeMillis();
        for (int i = 0; i < NUM_ITERATIONS; i++)
            fac = recursiveFac(FACT_N);
        duration = (System.currentTimeMillis() - timeBefore);
        println("Recursive: " + fac + " (" + duration + "ms)");

        timeBefore = System.currentTimeMillis();
        for (int i = 0; i < NUM_ITERATIONS; i++)
            fac = stirlingFac(FACT_N);
        duration = (System.currentTimeMillis() - timeBefore);
        println("Approximation: " + fac + " (" + duration + "ms)");

        timeBefore = System.currentTimeMillis();
        for (int i = 0; i < NUM_ITERATIONS; i++)
            fac = lookupTableFac(FACT_N);
        duration = (System.currentTimeMillis() - timeBefore);
        println("Lookup table with map: " + fac + " (" + duration + "ms)");

        timeBefore = System.currentTimeMillis();
        for (int i = 0; i < NUM_ITERATIONS; i++)
            fac = lookupTableArrayFac(FACT_N);
        duration = (System.currentTimeMillis() - timeBefore);
        println("Lookup table with array: " + fac + " (" + duration + "ms)");
    }
}

File: src-python/6-AlgoAnalysis/FastTest.py
import math
import time
import numpy as np

FACT_N = 20
NUM_ITERATIONS = 1_000_000

# Lookup Tabellen
lookup_table = {}
lookup_table_array = np.zeros(FACT_N + 1)


def iterative_fac(n):
    product = 1
    for k in range(1, n + 1):
        product *= k
    return product


def recursive_fac(n):
    if n == 0:
        return 1
    return n * recursive_fac(n - 1)


def stirling_fac(n):
    return round(math.sqrt(2 * math.pi * n) * (n / math.e) ** n)


def lookup_table_fac(n):
    if n not in lookup_table:
        lookup_table[n] = iterative_fac(n)
    return lookup_table[n]


def lookup_table_array_fac(n):
    if lookup_table_array[n] == 0:
        lookup_table_array[n] = iterative_fac(n)
    return lookup_table_array[n]


def run():
    print("Running factorial performance tests...\n")

    # Iterative
    start = time.time()
    for _ in range(NUM_ITERATIONS):
        fac = iterative_fac(FACT_N)
    duration = (time.time() - start) * 1000
    print(f"Iterative: {fac} ({int(duration)}ms)")

    # Recursive
    start = time.time()
    for _ in range(NUM_ITERATIONS):
        fac = recursive_fac(FACT_N)
    duration = (time.time() - start) * 1000
    print(f"Recursive: {fac} ({int(duration)}ms)")

    # Stirling
    start = time.time()
    for _ in range(NUM_ITERATIONS):
        fac = stirling_fac(FACT_N)
    duration = (time.time() - start) * 1000
    print(f"Approximation: {fac} ({int(duration)}ms)")

    # Lookup mit dict
    start = time.time()
    for _ in range(NUM_ITERATIONS):
        fac = lookup_table_fac(FACT_N)
    duration = (time.time() - start) * 1000
    print(f"Lookup table with map: {fac} ({int(duration)}ms)")

    # Lookup mit Array
    start = time.time()
    for _ in range(NUM_ITERATIONS):
        fac = lookup_table_array_fac(FACT_N)
    duration = (time.time() - start) * 1000
    print(f"Lookup table with numpy array: {int(fac)} ({int(duration)}ms)")


if __name__ == "__main__":
    run()